pkg/util/generics.go

```diff
@@ -8,3 +10,24 @@ func Contains[T comparable](s []T, e T) bool {
 	}
 	return false
 }
+
+type LazyOf[T any] struct {
+	New   func() (T, error)
+	once  sync.Once
+	value T
+}
+
+func (l *LazyOf[T]) Value() (T, error) {
+	var err error
+	if l.New != nil {
+		l.once.Do(func() {
+			l.value, err = l.New()
+			l.New = nil
+		})
+	}
+	return l.value, err
+}
+
+func NewLazyOf[T any](newfunc func() (T, error)) *LazyOf[T] {
+	return &LazyOf[T]{New: newfunc}
+}
```

Comment:
It's not completely clear to me why this lazy initialization is needed. Can't we just initialize when initializing the NodePool backend, or is it because the resources are not applied at that time?

The way to solve that would be to do the resolution whenever there is a need to lookup the CRD resources and not during the setup of the client. Since this is gonna be short lived I think we can keep it like this for now. But I would suggest to remove it long term to simplify the code.

Reply:
yes, it is because the CRDs are not there during the initialization of the backend.

I tried to delay the initialization of the resolver but it was a bit difficult to pass all the parameters across the functions all the way to where it is needed. I thought this was "simpler" 😄 

> But I would suggest to remove it long term to simplify the code.

yea, sure, we can do this